File: symfony/src/Service/Tournaments/AwardedPrizeService.php
<?php

namespace App\Service\Tournaments;

use App\Entity\Tournaments\AwardedPrize;
use App\Repository\Tournaments\AwardedPrizeRepository;

class AwardedPrizeService
{
    public function __construct(
        private AwardedPrizeRepository $repository,
    ) {}

    public function create(array $data): AwardedPrize
    {
        throw new \LogicException('Not implemented');
    }

    public function update(AwardedPrize $entity, array $data): AwardedPrize
    {
        throw new \LogicException('Not implemented');
    }

    public function claim(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('AwardedPrize not found: ' . $id);
        $entity->claim();
        $this->repository->save($entity, flush: true);
    }

    public function setClaimed(int $id, string $value): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('AwardedPrize not found: ' . $id);
        $entity->setClaimed($value);
        if ($value === 'TRUE') {
            $entity->claim(); // @on(claimed = true)
        }
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/tournaments/awarded_prize_service.rb
module Tournaments
  class AwardedPrizeService
    def initialize(repository = AwardedPrize)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def claim(id)
      instance = AwardedPrize.find(id)
      instance.claim()
      instance.save!
    end

    # triggered by @on(claimed = true)
    def set_claimed(id, value)
      instance = AwardedPrize.find(id)
      instance.claimed = value
      if value.to_s.upcase == 'TRUE'
        instance.claim
      end
      instance.save!
    end
  end
end
